substra/sdk/backends/local/backend.py

```diff
@@ -757,12 +759,14 @@ def _add_composite_traintuple(
             key=key,
             creator=_BACKEND_ID,
             algo={
-                "hash": spec.algo_key,
+                "key": spec.algo_key,
+                "hash": algo.content.hash_,
                 "name": algo.name,
                 "storage_address": algo.content.storage_address
             },
             dataset={
-                "opener_hash": spec.data_manager_key,
+                "key": spec.data_manager_key,
+                "opener_hash": spec.data_manager_key,  # TODO: pass actual opener hash
```

Comment:
l.700, we get the dataset for existence validation `self._db.get(schemas.Type.Dataset, spec.data_manager_key)`,
so you can replace it by `dataset = self._db.get(schemas.Type.Dataset, spec.data_manager_key)` and do for the opener hash: 
```python
"opener_hash": dataset.opener.hash_
```